src/Avalonia.Base/Data/Core/BindingExpression.cs

```diff
@@ -231,10 +233,14 @@ internal void OnNodeValueChanged(int nodeIndex, object? value, Exception? dataVa
 
         if (nodeIndex == _nodes.Count - 1)
         {
-            // If the binding source is a data context without any path and is currently null, treat it as an invalid
-            // value. This allows bindings to DataContext and DataContext.Property to share the same behavior.
-            if (value is null && _nodes[nodeIndex] is DataContextNodeBase)
-                value = AvaloniaProperty.UnsetValue;
+            if (_mode == BindingMode.OneTime)
+            {
+                // In OneTime mode, only changing the data context updates the binding.
+                if (!_shouldUpdateOneTimeBindingTarget && _nodes[nodeIndex] is not DataContextNodeBase)
```

Comment:
Is this data context only, or also any source changes?
For example, if I set relative source to TemplatedParent, and then move control to another template.
Or set DynamicResource as a source and change it, without any path.

Reply:
I've just checked WPF: `RelativeSource` bindings never update, _even in OneWay_ mode. I've tried `AncestorType` and `TemplatedParent` then moving controls from one container to another. And a `DynamicResource` can only be applied to a `DependencyProperty`.

So we should choose a behavior for Avalonia here. It's a pretty niche use case, I'd say we keep the documented behavior of _once per DataContext_, not _once per source_.